### ISSUE TITLE
Intermittent instability (binds not corresponding, query strings being intermixed)

### ISSUE THREAD

**Issue by mattadams** [flightaware/Pgtcl#43]
When I was using Pgtcl release 2.7.5 everything was quite stable, however I have upgraded to 2.8.0 (comes out as 2.7.7 when compiled) and am encountering intermittent instability.

With 2.8.0, I will get curious errors like this:

```
postgres[73105]: [9-1] 2022-04-13 19:40:12.094 MDT [73105] ERROR:  bind message supplies 3 parameters, but prepared statement "" requires 7
postgres[73105]: [9-2] 2022-04-13 19:40:12.094 MDT [73105] STATEMENT:  
postgres[73105]: [9-3]            INSERT INTO logs (inserted_at, severity, log, creator, pid, tid, queue_id)
postgres[73105]: [9-4]            VALUES ($1, $2, $3, $4, $5, $6, $7)
postgres[73105]: [9-5]
```

and

```
postgres[73111]: [9-1] 2022-04-13 19:40:12.750 MDT [73111] ERROR:  syntax error at or near "NULextended" at character 88
postgres[73111]: [9-2] 2022-04-13 19:40:12.750 MDT [73111] STATEMENT:  UPDATE datasets SET imported_at = refreshed_at WHERE code = 'ABST' AND (imported_at IS NULextended ->> 'event' = 'appended_at'
postgres[73111]: [9-3]                
```

In the latter case it looks like two separate SQL statements are being intermixed because the actual SQL for the latter comes from:

```
UPDATE datasets SET ${event}_at = refreshed_at WHERE code = [::pg::quote $dataset_code] AND (${event}_at IS NULL OR refreshed_at > ${event}_at) RETURNING refreshed_at, ${event}_at
```

In this case the actual SQL error makes absolutely no sense, because the bit that reads ```extended ->> 'event' = 'appended_at'``` comes from another query entirely.

Thoughts?

**Comment by mattadams** [comment]
Both of these statements are run through ```pg::sqlexec``` (via ```Pg_exec``` in ```pgtclCmds.c```).

**Comment by resuna** [comment]
Can you provide the source code that is generating this SQL?

**Comment by mattadams** [comment]
I will do my best to provide something that replicates it. Unfortunately it is very difficult to reproduce reliably in my environment. Either that, or I'll isolate it myself and provide a patch.

I did see ticket [1bd796d9c2](https://core.tcl-lang.org/tcl/tktview?name=1bd796d9c2) in Tcl 8.6.12 and I am hoping that the problem doesn't stem from this.

**Comment by resuna** [comment]
Can you at least provide the Tcl code that generated the errors you listed so I can get an idea of where to look?

The possibility of an underlying bug in Tcl_DString is scary.

**Comment by resuna** [comment]
Or the parameters to sqlexec?

**Comment by resuna** [comment]
@mattadams ^

**Comment by resuna** [comment]
The empty string in the bind error is apparently a PostgreSQL artifact. Which is a pity because I was hoping it would be a useful diagnostic. This stackoverflow article shows it happening with javascript bindings. :(

https://stackoverflow.com/questions/43755454/bind-message-supplies-1-parameters-but-prepared-statement-requires-2

**Comment by resuna** [comment]
@maddadams Try branch BCK-6946

**Comment by resuna** [comment]
Fixed in 3.1.0
